lib/install/management/configure.go

```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
+		d.deleteISOs(ds, settings)
+		d.deleteSnapshot(newSnapshotRef, snapshotName, conf.Name)
+
 		return err
 	}
 
-	err = d.update(conf, settings, isConfigureOp)
+	// if we are upgrading evaluate need for inventory upgrade
+	if d.Action == UpgradeAction && d.session.IsVC() {
+		err := d.updateInventory(conf.Name)
+		if err != nil {
+			// if we returned an error then no Inventory work was
+			// started.  If we failed during update we warn the user.
+			d.op.Errorf("An error prevented inventory changes: %s", err)
+		}
+	}
 
 	// If successful try to grant permissions to the ops-user
-	if err == nil && conf.ShouldGrantPerms() {
-		err = opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+	if conf.ShouldGrantPerms() {
+		// TODO: In a DRS environment that's not been upgraded to 1.4
+		// this will attempt to apply permissions to the VmFolder. Does
+		// that work & do we really want that?
+		err := opsuser.GrantOpsUserPerms(d.op, d.session, conf)
 		if err != nil {
-			// Update error message and fall through to roll back
 			err = errors.Errorf("Failed to grant permissions to ops-user, failure: %s", err)
 		}
 	}
 
+	// compatible with old version's upgrade snapshot name
+	if oldSnapshot != nil && (vm.IsConfigureSnapshot(oldSnapshot, ConfigurePrefix) || vm.IsConfigureSnapshot(oldSnapshot, UpgradePrefix)) {
+		d.deleteSnapshot(&oldSnapshot.Snapshot, oldSnapshot.Name, conf.Name)
+	}
+
+	return nil
+}
+
+// updateInventory updates VCH Folder inventories.
+//
+// If the VCH resides in the VMFolder then a new folder will
+// be created and the VCH and all Containers will be moved into the
+// new folder.  If the VCH already resides in a folder other than
+// the VMFolder, then no action will be taken.
+//
+// vApp support planned: https://github.com/vmware/vic/issues/7670
+func (d *Dispatcher) updateInventory(name string) error {
+	defer trace.End(trace.Begin("", d.op))
+
+	appName := d.appliance.Name()
+
+	d.op.Infof("Updating inventory for %s or %s", name, appName)
+
+	// get the current appliance folder
+	currentFolder, err := d.appliance.Folder(d.op)
 	if err != nil {
-		// Roll back
-		d.op.Errorf("Failed to upgrade: %s", err)
-		d.op.Infof("Rolling back upgrade")
+		d.op.Errorf("Unable to get current folder:  %s", err)
+		return err
+	}
 
-		if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
-			d.op.Errorf("Failed to revert appliance to snapshot: %s", rerr)
-			return err
-		}
-		d.op.Infof("Appliance is rolled back to old version")
+	// vApp support planned: https://github.com/vmware/vic/issues/7670
+	if currentFolder.Reference() != d.session.VMFolder.Reference() || d.vchPool.Reference().Type == "VirtualApp" {
+		d.op.Debugf("No inventory update completed for %s", name)
+		return nil
+	}
 
-		d.deleteUpgradeImages(ds, settings)
-		d.retryDeleteSnapshotByRef(newSnapshotRef, snapshotName, conf.Name)
+	// concatenate w/vch name so we can identify if failure
+	tempName := fmt.Sprintf("%s-%s", name, uid.New().Truncate().String())
```

Comment:
We should ensure that this doesn't violate the name length limits. It may be possible to use the name convention logic to do so (`lib/portlayer/util/util.go`).